### PR TITLE
Detect and serialize objects with `null` prototype; provide `shouldJSONSerializeBody` hook to override behavior

### LINKED ISSUE #67: Evaluate AS PR 5070 (null prototype issue)
We closed https://github.com/apollographql/apollo-server/pull/5070 to move it to this repository. We should look to see if the change is still applicable.

### LINKED ISSUE #68: RESTDataSource - cannot convert object to primitive
Hi,
I have question about RESTDataSource docs: 
https://www.apollographql.com/docs/apollo-server/features/data-sources.html#HTTP-Methods

What have you passed as a movie in post/put/patch requests? Because it seems that you can't pass object there. You should JSON.strigify it first. Otherwise you will get "cannot convert object to primitive" from node-fetch. Am I right?

Here is my data source:

```
export default class DataSource extends BaseDataSource {
    uri = '/travelSeasons';

    async getTravelSeasons() {
        return this.get(this.uri);
    }

    async addTravelSeason(travelSeason: TravelSeason) {
        return this.post<TravelSeason>(this.uri, travelSeason);
    }

    async updateTravelSeason(travelSeason: TravelSeason) {
        return this.patch<TravelSeason>(this.uri + `/${travelSeason.id}`, travelSeason);
    }
}

```
And when I comment lines below:
```
export default class BaseDataSource extends RESTDataSource {
    baseURL = config.API_URL;

    // protected post<TResult = any>(path: string, body?: Body, init?): Promise<TResult> {
    //     return super.post(path, JSON.stringify(body), {
    //         headers: {
    //             "Content-Type": "application/json"
    //         },
    //         ...init
    //     });
    // }

    ...
}
```

I will get this:

```
{
  "data": {
    "addTravelSeason": null
  },
  "errors": [
    {
      "message": "Cannot convert object to primitive value",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "addTravelSeason"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "stacktrace": [
            "TypeError: Cannot convert object to primitive value",
            "    at String (<anonymous>)",
            "    at Request.Body (/usr/src/app/node_modules/node-fetch/lib/index.js:172:10)",
            "    at new Request (/usr/src/app/node_modules/node-fetch/lib/index.js:1169:8)",
            "    at DataSource.<anonymous> (/usr/src/app/node_modules/apollo-datasource-rest/dist/RESTDataSource.js:140:29)",
            "    at Generator.next (<anonymous>)",
            "    at fulfilled (/usr/src/app/node_modules/apollo-datasource-rest/dist/RESTDataSource.js:4:58)",
            "    at <anonymous>",
            "    at process._tickCallback (internal/process/next_tick.js:188:7)"
          ]
        }
      }
    }
  ]
}
```
My node version is: 8.11.3, all libraries are up to date.

Btw what about "Content-Type": "application/json" header? Is there a good way to add it to my datasource by default (in base class or sth - I guess some kind of adapter) ? Is my solution ok?

### PR DESCRIPTION
**Pull request by trevor-scheer**
This PR refines the body serialization conditional to be more straightforward (rather than a confusing mix of do's and don'ts). Bodies are JSON serialized if:
* It's an array or a plain object
* It has a `.toJSON()` function on it (with the exception of `Buffer` which we do not want to stringify)

Lodash utilities are widely used and well-tested, so I went with the `isPlainObject` implementation they provide in order to correctly detect objects. The existing conditional falls short for objects with a `null` prototype.

For users desiring to override this behavior, we've provided the new protected method `shouldJSONSerializeBody`. This method should return a `boolean` in order to tell RESTDataSource whether or not to call `JSON.stringify` on the body in question.

Some implementations of `fetch` support different data types (such as `FormData`). You're welcome to use those, but it's your responsibility to use libraries that are compatible with each other, i.e. `node-fetch` is compatible with `form-data`.

Fixes #67
Fixes #68
Ref: https://github.com/apollographql/apollo-server/pull/5070

### CODE REVIEW

**Comment by aeddie-zapidhire** [comment]
Looks good to me. Only comment I might possibly make is that https://github.com/sindresorhus/is-plain-obj could be a smidge cleaner and more performant.

**Comment by glasser** [comment]
What about taking the opposite approach: serialize anything truthy that is *not* a string or Buffer?

**Comment by glasser** [comment]
(That's a good fit for our FetcherRequestInit `body?: string | Buffer`: anything that can't go directly here (ie, not string or Buffer or undefined) gets treated as JSON. So probably "not undefined" rather than "truthy" in my previous suggestion.)

**Comment by trevor-scheer** [comment]
@glasser that change will fail one of the existing tests (that I fixed in this PR):
`does not serialize a request body that is not an object`

There's some additional context, particularly in these 2:
https://github.com/apollographql/apollo-server/pull/1316
https://github.com/apollographql/apollo-server/pull/5070

So...if you still think we should go that route I think we'll to add some error handling around the stringify call and possibly add a hook for handling those errors?

**Comment by glasser** [comment]
Hmm. Should we consider just explicitly supporting form-data in Fetcher (even though the Apollo Server/Gateway use cases don't need it)? Would require us to have a reasonably portable way of describing a FormData in TypeScript and a reasonable level of confidence that all the important implementations support it with portability between FormData implementations. (My vague memory is of looking into this at some point and deciding that they do mostly support it.)

It just seems odd to introduce a bunch of complexity just in order to ensure we can pass a FormData to an API that is defined as only taking string or Buffer.

File: src/RESTDataSource.ts [hunk 1]
```diff
@@ -208,6 +209,25 @@ export abstract class RESTDataSource {
     }
   }
 
+  protected shouldJSONSerializeBody(body: RequestWithBody['body']): boolean {
+    return !!(
+      // We accept arbitrary objects and arrays as body and serialize them as JSON.
+      (
+        Array.isArray(body) ||
+        isPlainObject(body) ||
```
Comment (glasser) [inline src/RESTDataSource.ts:217]: i still kinda feel like "anything but string or Buffer" or "anything but string, Buffer, or FormData" would be a better default — the goal of this is to produce something that the Fetch API understands, and so it feels like we should serialize anything other than what Fetch understands.

On the other hand, real Fetch APIs also take other stuff like readable streams, URLSearchParameters, TypedArray, Blob, etc, and perhaps your choice here is better for those things (and we're already moving away from "serialize anything that our Fetcher interface doesn't expect" due to FormData).

So I guess this does seem fine, given that it's just a default people can change.

File: src/RESTDataSource.ts [hunk 2]
```diff
@@ -208,6 +209,25 @@ export abstract class RESTDataSource {
...
+        // We serialize any objects that have a toJSON method (except Buffers or things that look like FormData)
+        (body &&
+          typeof body === 'object' &&
+          'toJSON' in body &&
+          typeof (body as any).toJSON === 'function' &&
+          !(body instanceof Buffer) &&
+          // XXX this is a bit of a hacky check for FormData-like objects (in
+          // case a FormData implementation has a toJSON method on it)
+          (body as any).constructor.name !== 'FormData')
```
Comment (glasser) [inline src/RESTDataSource.ts:226]: do we want that to be `constructor?.name`?

**Review by glasser — APPROVED** [review]

**Pull request merged**